### DISCUSSION TITLE
Availability of chord-chart-studio source code

### DISCUSSION THREAD

**Discussion by bbinet** [no-chris/chord-mark#550] [Q&A] [answered]
I was wondering if the source code for chord-chart-studio was available?

By the way, chord-chart-studio is awesome to edit chords!

**Comment by no-chris** [comment]
I need to prepare it a bit but that could be possible I guess. May I ask why you are interested, are you looking to contribute?

Thanks for your interest and your comments

**Comment by bbinet** [comment]
I often want to look at the source code when using software that I like (and I prefer to use opensource software).
I may be interested to contribute some features, for example:
- I'd like to have the possibility to export PDF including multiples songs to get a songbook (with or without chords).
- I would also like to be able to import songs in batch
- It could also be useful to have a tagging system so that we can easily organize songs, to make them more easily searchable
I don't know yet if I will be able to contribute as I need to see the source code first, but at least I'm interested!

**Comment by no-chris** [comment]
Thanks for the feedback and the features ideas 🙌🏻 

Tagging and organising songs would indeed be one the next logical steps for Chord Chart Studio. We've discussed metadata in an other thread, but the possibility to arrange songs in collections (and exporting them in a single PDF) is also a must have!
I haven't thought of bulk import, that's interesting also

I'll can add you to the private repo if you want to have a look

**Comment by bluestrat404** [comment]
Hey @no-chris, first, AWESOME project. Second, could I be added to the private repo? I'd like to look into adding two things:

1. A formatter for the PDF to create "Themes", mostly for an output like the PDF attached below, for readability. I already have the CSS working on my box to create these PDFs, so I'm hoping this could be a relatively easy PR.

2. Stretch goal: a way to output very simple piano score PDFs for pianists not used to playing from chord charts, probably using LilyPond.

Let me know if either of these sounds interesting. Either way, kudos on the app and the ChordMark format!

Thanks

[OComeOComeEmmanuelEmCapo8Cm.pdf](https://github.com/no-chris/chord-mark/files/10305199/OComeOComeEmmanuelEmCapo8Cm.pdf)

**Reply by no-chris** [reply]
Nice rendering! 

1/ would be a cool feature, indeed.

I would suggest the following:
- add your theme to https://github.com/no-chris/chord-mark/tree/master/packages/chord-mark-themes 
- create an new option in the print mode of Chord Chart Studio to apply the selected theme, those are good places to look at:
  - https://github.com/no-chris/chord-chart-studio/blob/master/packages/chord-chart-studio/src/optionsPanels/rendering/allWidgets.js
  - https://github.com/no-chris/chord-chart-studio/blob/master/packages/chord-chart-studio/src/db/options/editorModeOptions.js
  - https://github.com/no-chris/chord-chart-studio/blob/c9492adcc340a8cd093bfb743c628b5aaf017b3e/packages/chord-chart-studio/src/songRenderers/printPreview/_components/PrintPreview.js#L30
  - https://github.com/no-chris/chord-chart-studio/blob/master/packages/chord-chart-studio/src/songRenderers/printPreview/_containers/PrintPreview.js
 
The application uses React/Redux patterns that are I think quite common, but let me know if you need help navigating the code

Regarding 2/ if I understand well you would like to actually render the notes in a chord for pianist, using Lilypond?

Thanks for your interest 🙏🏻 

Christophe

**Reply by bluestrat404** [reply]
Thanks for adding me and for pointing me in the right direction. I'm excited to take a look.

Regarding #2, that is correct: I'd like to render the notes and durations of each chord on a staff, maybe a grand staff with just the 1 and 5 in the bass, and all the notes in the treble — just enough to give classically trained pianists a start — and then print the lyrics too. Since ChordMark requires the user to input chord durations (unlike ChordPro), this should be possible. As far as using Lilypond, that seems like the simplest way to do it that I'm aware of, but I am certainly open to other suggestions.

I hope to start working on this sometime this week and will reach out with any questions.

Adam

**Reply by no-chris** [reply]
> Thanks for adding me and for pointing me in the right direction. I'm excited to take a look.

🙌🏻 

> Regarding https://github.com/no-chris/chord-mark/pull/2, that is correct: I'd like to render the notes and durations of each chord on a staff, maybe a grand staff with just the 1 and 5 in the bass, and all the notes in the treble — just enough to give classically trained pianists a start — and then print the lyrics too. Since ChordMark requires the user to input chord durations (unlike ChordPro), this should be possible. As far as using Lilypond, that seems like the simplest way to do it that I'm aware of, but I am certainly open to other suggestions.

Very interesting. Another way of doing it could be to write a ChordMark => MusicXML converter, maybe that could cover the use case with even better interoperability. I've played a bit with MusicXML in the past while writing a ChordSymbol plugin for an iRealB => MusicXML converter: 
- https://github.com/infojunkie/ireal-musicxml
- https://github.com/no-chris/chord-symbol-musicxml

It seems that there are some libraries allowing to render MusicXML directly in the browser https://blog.karimratib.me/demos/musicxml/

**Comment by bluestrat404** [comment]
That sounds good. It looks like we could go from ChordMark to MusicXML and then display/print using OpenSheetMusicDisplay, keeping everything in the JS world and easily hosted as a static site like now.

I’ll work on the theme thing and then this.

**Comment by no-chris** [comment, marked as answer]
Chord charts studio is now GPL3
https://github.com/no-chris/chord-chart-studio
